odl/util/statistics.py

```diff
@@ -0,0 +1,147 @@
+﻿# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Utilities for computing statistics on images."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+import numpy as np
+
+
+__all__ = ('mse', 'psnr', 'estimate_noise')
+
+
+def mse(true, noisy):
+    """Return the Mean Squared Errror.
+
+    Parameters
+    ----------
+    true : array-like
+    noisy : array-like
+
+    Returns
+    -------
+    mse : float
+
+    Examples
+    --------
+    >>> true = [1, 1, 1, 1, 1]
+    >>> noisy = [1, 1, 1, 1, 2]
+    >>> result = mse(true, noisy)
+    >>> print('{:.3f}'.format(result))
+    0.200
+    """
+    true = np.asarray(true)
+    noisy = np.asarray(noisy)
+    return np.mean((true - noisy) ** 2)
+
+
+def psnr(true, noisy, normalize=False):
+    """Return the Peak signal-to-noise ratio.
+
+    Parameters
+    ----------
+    true : array-like
+    noisy : array-like
+    normalize : bool
+        If true, normalizes the true and noisy signal to have the same mean
+        and variance before comparison.
+
+    Returns
+    -------
+    psnr : float
+
+    Examples
+    --------
+    >>> true = [1, 1, 1, 1, 1]
+    >>> noisy = [1, 1, 1, 1, 2]
+    >>> result = psnr(true, noisy)
+    >>> print('{:.3f}'.format(result))
+    6.990
+
+    If true == noisy, the result is infinite
+
+    >>> psnr([1, 1], [1, 1])
+    inf
+
+    If `true == 0` but `noisy != 0`, the result is negative infinity
+
+    >>> psnr(0, 1)
+    -inf
+    """
+    true = np.asarray(true)
+    noisy = np.asarray(noisy)
+
+    if normalize:
+        noisy = noisy + np.mean(true - noisy)
+        noisy = noisy * np.std(true) / np.std(noisy)
+
+    mse_result = mse(true, noisy)
+    max_true = np.max(np.abs(true))
+
+    if mse_result == 0:
+        return np.inf
+    elif max_true == 0:
+        return -np.inf
+    else:
+        return 20 * np.log10(max_true) - 10 * np.log10(mse_result)
+
+
+def estimate_noise(img):
+    """Estimate noise level in image.
```

Comment:
Does the reference say anything about the noise types for which it works? Maybe it's valid for more than white noise. 
What you're after here seems to be the `std` of the noise, given a number of samples drawn from distributions with same `std` but potentially different mean values, i.e. random variables `X_i = X + c_i` with the same `X` but different constants `c_i`. Additive noise for that matter. That rules out Poisson, but pink noise could work, too, for instance.
Perhaps name it `estimate_std` something?